packages/expo-modules-core/src/web/CoreModule.ts

```diff
@@ -33,23 +40,36 @@ class EventEmitter<TEventsMap extends EventsMap> implements EventEmitterType {
     eventName: EventName,
     listener: TEventsMap[EventName]
   ): void {
-    this.listeners?.get(eventName)?.delete(listener);
+    const hasRemovedListener = this.listeners?.get(eventName)?.delete(listener);
+    if (this.listenerCount(eventName) === 0 && hasRemovedListener) {
+      this.stopObserving(eventName);
+    }
   }
 
   removeAllListeners<EventName extends keyof TEventsMap>(eventName: EventName): void {
+    const listenerCount = this.listenerCount(eventName);
+
     this.listeners?.get(eventName)?.clear();
+    if (listenerCount > 0) {
+      this.stopObserving(eventName);
+    }
   }
 
   emit<EventName extends keyof TEventsMap>(
     eventName: EventName,
     ...args: Parameters<TEventsMap[EventName]>
   ): void {
-    this.listeners?.get(eventName)?.forEach((listener) => listener(...args));
+    const listeners = new Set(this.listeners?.get(eventName));
+    listeners.forEach((listener) => listener(...args));
```

Comment:
```suggestion
    const listeners = this.listeners?.get(eventName) ?? [];
    [...listeners].forEach((listener) => listener(...args));
```
You don't have to create a set here, but it's up to you ;)

Reply:
it's a set already, so it's probably cleaner to copy it instead of converting to array - wdyt?